### PR TITLE
compatibility with downloading data from gcp

### LINKED ISSUE #94: GCSFuse mount + Vertex AI custom training jobs support
## 🚀 Feature

DDP on models in GCP with data stored in GCS

### Question

Has `litdata` being tested in the case you want to train on GCP (Vertex AI) instead of just storing and streaming the data from GCS?

### Motivation

I've been trying to set up ddp on GCP (Vertex AI) using lighting Fabric with my data being stored in GCS, when you use GCS and Vertex AI you can "mount" a bucket to the instance(s)/containers running in their infrastructure using [GCSFuse](https://cloud.google.com/storage/docs/gcsfuse-integrations), in such case the only thing you need to do is replacing `gs://` by `/gcs/` and the bucket acts as a file system, does `litdata` have been tested under such setting? will it work ?

I tried already with `mosaicml-streaming` and I have ran into lots of throughput issues, that lead to data starvation in the multi-node setting

I wrote this cluster environment to use configure the cluster on vertex AI + Fabric:

```python
import os
import json

from lightning.fabric.plugins.environments.lightning import LightningEnvironment


class VertexAICluster(LightningEnvironment):
    """
    Configures distributed training on a vertex ai custom training job,
    ex:
        Consider a cluster with 3 nodes, each composed of 2 gpus

        The "cluster" key in CLUSTER_SPEC will be:
            {
                'workerpool0': ['cmle-training-workerpool0-d604929a6a-0:2222'],
                'workerpool1': [
                                'cmle-training-workerpool1-d604929a6a-0:2222',
                                'cmle-training-workerpool1-d604929a6a-1:2222'
                              ]
            }

        and each process scheduled will be under the "task" key, following the same example
        the three tasks will look like this:
            task0 ("chief" spawn process) -> node 0:
            {'type': 'workerpool0', 'index': 0}
            task 1 (on first node on workerpool1) -> node 1:
            {'type': 'workerpool1', 'index': 0}
            task 2 (on second node on workerpool1) -> node 2:
            {'type': 'workerpool1', 'index': 1}
    """

    def __init__(self):
        super().__init__()
        self.cluster_spec = json.loads(os.environ['CLUSTER_SPEC'])

    @property
    def main_address(self) -> str:
        return self.cluster_spec["cluster"]["workerpool0"][0].split(':')[0]

    @property
    def main_port(self) -> int:
        """Set common fixed MASTER_PORT port across processes
        """
        return int(self.cluster_spec["cluster"]["workerpool0"][0].split(':')[1])

    def node_rank(self) -> int:
        task = self.cluster_spec["task"]
        if task["type"] == "workerpool0":
            return 0
        else:
            return task["index"] + 1
```
do I need to set up some other `env` variables if I wanted to test `litdata`?, do the ones defined here in this cluster environment are used by `litdata`, in `mosaicml` you had to configure [these](https://docs.mosaicml.com/projects/streaming/en/stable/fundamentals/environments.html#)

### PR DESCRIPTION
**Pull request by dangthatsright**
Similar to the S3 data downloading, this works for GCP. I've tested it with my own data but not sure how I would set up tests for this repo.

The link to the contributor guideline is dead btw.

If there are specific places you would like me to update, please let me know.

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #94 #101

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/litdata/streaming/downloader.py [hunk 1]
```diff
@@ -87,6 +88,36 @@ def download_file(self, remote_filepath: str, local_filepath: str) -> None:
             pass
 
 
+class GCPDownloader(Downloader):
```
Comment (tchaton) [inline src/litdata/streaming/downloader.py:91]: Can you add a test where you mock the storage module ?

Comment (dangthatsright) [reply src/litdata/streaming/downloader.py:91]: added, lmk if that's what you want. i also see something in test/streaming/test_dataset.py breaking but i can't reproduce it. lmk if it's something i introduced that broke something

File: requirements.txt [hunk 2]
```diff
@@ -3,3 +3,4 @@ filelock
 numpy
 boto3[crt]
 requests
+google-cloud-storage
```
Comment (tchaton) [inline requirements.txt:6]: Can we move this to extras? This shouldn't be a default

File: tests/streaming/test_downloader.py [hunk 3]
```diff
@@ -13,6 +20,31 @@ def test_s3_downloader_fast(tmpdir, monkeypatch):
     popen_mock.wait.assert_called()
 
 
+def test_gcp_downloader(tmpdir, monkeypatch):
```
Comment (tchaton) [inline tests/streaming/test_downloader.py:23]: Looks good.

File: src/litdata/streaming/downloader.py (hunk 1 above)
Comment (tchaton) [reply src/litdata/streaming/downloader.py:91]: Looks good.

File: requirements.txt (hunk 2 above)
Comment (dangthatsright) [reply requirements.txt:6]: sounds good, i figured boto3 was here as well so i added it here

Comment (tchaton) [reply requirements.txt]: Yes, we will make it first class citizen when there is more demand for it

**Comment by tchaton** [comment]
Hey @dangthatsright. I have fixed the failure in the test. When you import an optional package, you need to make sure the logic is protected by a boolean to avoid failures.

I wondered if you would be keen on contributing the same feature for the `optimize` method, so users can generate their data and push directly to gcs.

Something like this:

```python

optimize(
    ....
    output_dir="gcs://....",
)

**Pull request merged**